### PR TITLE
Bring all disparate datetime handling into core helpers module.

### LINKED ISSUE #1024: Move all datetime handling into a shared core module
See https://github.com/GoogleCloudPlatform/gcloud-python/pull/1021#discussion_r35710351

We have (essentially) a novel approach to `datetime` in every sub-package, so we should get unified.

- From https://github.com/GoogleCloudPlatform/gcloud-python/pull/1044/files#r36545736: Hmm, maybe `_EPOCH` should be moved up to `gcloud._helpers`.

### PR DESCRIPTION
**Pull request by dhermes**
Fixes #1024.

This is best reviewed commit by commit. I'm happy to break it into more PRs if desired.

### CODE REVIEW

File: gcloud/_helpers.py [hunk 1]
```diff
@@ -207,7 +209,66 @@ def _determine_default_project(project=None):
...
+    """Convert a zone-aware datetime to integer milliseconds.
+
+    :type when: :class:`datetime.datetime`
+    :param when: the datetime to convert
+
+    :rtype: integer
+    :returns: milliseconds since epoch for ``when``
+    """
+    micros = _microseconds_from_datetime(when)
+    millis, _ = divmod(micros, 1000)
```
Comment (tseaver) [inline gcloud/_helpers.py]: Seems like `millis = micros // 1000` would be simpler.

Comment (dhermes) [inline gcloud/_helpers.py]: SGTM

**Comment by tseaver** [comment]
Other than the one quibble, LGTM

**Comment by dhermes** [comment]
Addressed `divmod` there and elsewhere (all added in this PR, and removed in the final commit).

**Pull request merged**